### PR TITLE
156 spotify   generating refresh tokens

### PR DESCRIPTION
**Pull request by BitsyBirb**
# What type of PR is this? (Check all that apply)

- [X] ✨**feature**: Introduces completely new code or new features.
- [ ] 🐛**fix**: Implements changes that fix a bug. Ideally, reference an issue if present.
- [ ] ♻️**refactor**: Includes any code-related change that is neither a fix nor a feature.
- [ ] ✅**build**: Encompasses all changes related to the build of the software, including changes to dependencies or the addition of new ones.
- [ ] ⚡️**test**: Pertains to all changes regarding tests, whether adding new tests or modifying existing ones.
- [ ] 🚰**ci**: Involves all changes related to the configuration of continuous integration, such as GitHub Actions or other CI systems.
- [ ] 📚**docs**: Includes all changes to documentation, such as README files, or any other documentation present in the repository.
- [ ] 🗑️**chore**: Captures all changes to the repository that do not fit into the above categories.

# Description

What did you change? How did you change it?
My main changes were in the newly created file:
1) `~/apps/expo/src/components/spotify/spotifyToken.tsx`
This file is the most important. It implements authentication via Auth-Session which uses OAuth2 allowing the user to approve and authenticate Barbell to access their Spotify data via the Spotify API.

After the user authenticates the app the initial tokens are generated to further refresh access tokens (as they have a time limit of an hour) when necessary through the refresh token generated.

The access token is checked whenever the updateSpotifyData() function is called (which is the main function to be exported and used by the rest of the team whenever the time comes for the frontend team).

Though it works in isolation, when it comes to more involved usage/use-cases there are a few things to address but those should be handled by the calling files.

One last note was that there are three main functions that can (and should be exported) are:
loginSpotifyButton() : A button to prompt the OAuth login session.

logoutSpotify(): Not a button, but can be added to a button. Just clears tokens and auth code to allow reprompt of login button.

updateTrackData(userID: integer): Pass the userID which should be updated from the running instance of the app. Should ONLY be that of the user which authenticated the app. This is the most important function as it will be used to update data in the database for that user (which is queried and used for all other display stuff like the Spotify widget to be implemented).

# Tests

How was this tested?

- [ ] Unit tests
- [ ] Integration tests
- [ ] E2E tests
- [X] Manual tests
- [ ] Tests were NOT needed
- [ ] Other (explain below)

# [Optional] Screenshots
Widget on Discord:
image

API Requests, token refresh, etc:
image

Output on frontend of Barbell (the time is inconsistent as it took me a bit to take the screenshot, ignore that):
image

# Documentation

- [ ] Added to README.me
- [ ] Seperate document
- [X] NO documentation needed

Link to external documentation: It's all linked within the files themselves. Refer to file comments.

# [Optional] Are there any post-deployment tasks we need to perform?

Though not post-deployment, there are a few known bugs that I have had issues trying to fix.
1) Issues on `~/apps/expo/src/app/spotify/index.tsx` regarding infinite looping of updateSpotifyData(). Function works normally but for some reason that page keeps invoking it causing the database to have a ton of mutation requests.

2) Need to add a dynamic userID on the above file using the context Marcos implemented a while back. Not relevant as much for this PR as that was just the page being used to test the overall functionality of our Spotify files, but in the end this will be 

3) We should find a way to move the client_id and client_secret (contents of `~/apps/expo/src/components/spotify/spotifyCreds.tsx`) into a .env file to keep it secret and out of main branch.

### CODE REVIEW

File: apps/expo/src/components/spotify/spotifyCreds.tsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { makeRedirectUri } from "expo-auth-session"
```
Comment (BitsyBirb) [inline apps/expo/src/components/spotify/spotifyCreds.tsx:1]: This is the file we should aim to hide using a .env file.

**Comment by BitsyBirb** [comment]
Make sure to run it in dev mode if possible. Forgot to mention that.

**Comment by StefanoRubini** [comment]
I have taken a look at the files changed and cannot seem to find anything inherently wrong with your code. I appreciate your efforts for leaving comments that describe what is being done! 😃

**Review by StefanoRubini — COMMENTED** [review]
I have taken a look at the files changed and cannot seem to find anything inherently wrong with your code. I appreciate your efforts for leaving comments that describe what is being done! 😃

**Review by supakeshi — CHANGES_REQUESTED** [review]
Please add usage comments so we can use/test your API

**Review by thienguen — APPROVED** [review]

**Comment by thienguen** [comment]
> Please add usage comments so we can use/test your API

its lie in the Spotify route, usage is to retrieve the current playing song on Spotify of the current log-in users, 

the matter of making it real-time depends on the WebSocket.

**Pull request merged**